File: python/python/openapp_sdk/resources/public_access.py
"""`Public Access` sub-client.

These endpoints run on opaque session / invite tokens embedded in the URL, not
on the OpenApp API key. They are exposed here so tools can drive them
programmatically (e.g. a concierge script that opens a door for a delivery
driver via a public invite).
"""

from __future__ import annotations

from typing import Any

from ._base import _BaseResource


class PublicAccessClient(_BaseResource):
    # -- Invites -------------------------------------------------------------

    async def get_invite(self, invite_token: str) -> dict[str, Any]:
        return await self._client._request("GET", f"/public/access/invites/{invite_token}")

    async def claim_invite(self, invite_token: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/invites/{invite_token}/claim", body=body
        )

    async def execute_invite(self, invite_token: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/invites/{invite_token}/execute", body=body
        )

    async def start_invite_session(self, invite_token: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/invites/{invite_token}/session", body=body
        )

    # -- Portals -------------------------------------------------------------

    async def get_portal(self, public_portal_id: str) -> dict[str, Any]:
        return await self._client._request("GET", f"/public/access/portals/{public_portal_id}")

    async def portal_lights(self, public_portal_id: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/portals/{public_portal_id}/lights", body=body
        )

    async def portal_open(self, public_portal_id: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/portals/{public_portal_id}/open", body=body
        )

    async def portal_reachable(self, public_portal_id: str) -> dict[str, Any]:
        return await self._client._request(
            "GET", f"/public/access/portals/{public_portal_id}/reachable"
        )

    async def portal_start_session(self, public_portal_id: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/portals/{public_portal_id}/sessions", body=body
        )

    async def portal_targets(self, public_portal_id: str) -> dict[str, Any]:
        return await self._client._request(
            "GET", f"/public/access/portals/{public_portal_id}/targets"
        )

    # -- Sessions ------------------------------------------------------------

    async def get_session(self, session_id: str) -> dict[str, Any]:
        return await self._client._request("GET", f"/public/access/sessions/{session_id}")

    async def cancel_session(self, session_id: str) -> dict[str, Any]:
        return await self._client._request("POST", f"/public/access/sessions/{session_id}/cancel")

    async def decline_session(self, session_id: str) -> dict[str, Any]:
        return await self._client._request("POST", f"/public/access/sessions/{session_id}/decline")

    async def session_lights(self, session_id: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/sessions/{session_id}/lights", body=body
        )

    async def session_notify_message(self, session_id: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/sessions/{session_id}/notify-message", body=body
        )

    async def session_open(self, session_id: str, **body: Any) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/public/access/sessions/{session_id}/open", body=body
        )

    async def session_streams(self, session_id: str) -> dict[str, Any]:
        return await self._client._request("GET", f"/public/access/sessions/{session_id}/streams")

File: rust/crates/core/src/resources/public_access.rs
//! `Public Access` resource group.
//!
//! These endpoints are authentication-free (they run on opaque session / invite
//! tokens in the path) but are still exposed through the SDK so tooling can drive
//! them programmatically.

use std::sync::Arc;

use reqwest::Method;

use super::JsonValue;
use crate::{
    error::SdkError,
    transport::{RequestSpec, Transport},
};

#[derive(Debug, Clone)]
pub struct PublicAccessClient {
    transport: Arc<Transport>,
}

impl PublicAccessClient {
    pub(crate) fn new(transport: Arc<Transport>) -> Self {
        Self { transport }
    }

    // -- Invites -----------------------------------------------------------------

    pub async fn get_invite(&self, invite_token: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/invites/{invite_token}");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::GET,
                path: &path,
                ..Default::default()
            })
            .await
    }

    pub async fn claim_invite(
        &self,
        invite_token: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/invites/{invite_token}/claim");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn execute_invite(
        &self,
        invite_token: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/invites/{invite_token}/execute");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn start_invite_session(
        &self,
        invite_token: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/invites/{invite_token}/session");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    // -- Portals -----------------------------------------------------------------

    pub async fn get_portal(&self, public_portal_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/portals/{public_portal_id}");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::GET,
                path: &path,
                ..Default::default()
            })
            .await
    }

    pub async fn portal_lights(
        &self,
        public_portal_id: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/portals/{public_portal_id}/lights");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn portal_open(
        &self,
        public_portal_id: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/portals/{public_portal_id}/open");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn portal_reachable(&self, public_portal_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/portals/{public_portal_id}/reachable");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::GET,
                path: &path,
                ..Default::default()
            })
            .await
    }

    pub async fn portal_start_session(
        &self,
        public_portal_id: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/portals/{public_portal_id}/sessions");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn portal_targets(&self, public_portal_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/portals/{public_portal_id}/targets");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::GET,
                path: &path,
                ..Default::default()
            })
            .await
    }

    // -- Sessions ----------------------------------------------------------------

    pub async fn get_session(&self, session_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::GET,
                path: &path,
                ..Default::default()
            })
            .await
    }

    pub async fn cancel_session(&self, session_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}/cancel");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                ..Default::default()
            })
            .await
    }

    pub async fn decline_session(&self, session_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}/decline");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                ..Default::default()
            })
            .await
    }

    pub async fn session_lights(
        &self,
        session_id: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}/lights");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn session_notify_message(
        &self,
        session_id: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}/notify-message");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn session_open(
        &self,
        session_id: &str,
        body: &JsonValue,
    ) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}/open");
        self.transport
            .request_json::<JsonValue, JsonValue>(RequestSpec {
                method: Method::POST,
                path: &path,
                body: Some(body),
                ..Default::default()
            })
            .await
    }

    pub async fn session_streams(&self, session_id: &str) -> Result<JsonValue, SdkError> {
        let path = format!("/public/access/sessions/{session_id}/streams");
        self.transport
            .request_json::<(), JsonValue>(RequestSpec {
                method: Method::GET,
                path: &path,
                ..Default::default()
            })
            .await
    }
}
